boards/BMS/Src/Io/Io_LTC6813.c

```diff
@@ -0,0 +1,368 @@
+ */
+static ExitCode Io_LTC6813_EnterReadyState(void);
+
+/**
+ * Start ADC conversions for all LTC6813 chips on the daisy chain.
+ * @return EXIT_CODE_OK if the command used to toggle the ADC can be sent to the
+ * device without timing out or errors. Else, EXIT_CODE_UNIMPLEMENTED.
+ */
+static ExitCode Io_LTC6813_StartADCConversion(void);
+
+/**
+ * Check if all LTC6813 chips in the daisy chain have all completed converting analogue
+ * cell voltages to digital voltages.
+ * @return EXIT_CODE_OK if all LTC6813 chips on the daisy chain have completed ADC
+ * conversions. EXIT_CODE_TIMEOUT if ADC conversions could not be completed
+ * before timing out. EXIT_CODE_UNIMPLEMENTED if the command sent and received
+ * to check the status of ADC conversions was not transmitted or received
+ * successfully.
+ */
+static ExitCode Io_LTC6813_PollAdcConversion(void);
+
+static void Io_LTC6813_ParseCellsAndPerformPec15Check(
+    size_t   current_ic,
+    size_t   current_register_group,
+    uint8_t *rx_cell_voltages);
+
+static uint16_t Io_CalculatePec15(uint8_t *data, uint32_t size)
+{
+    size_t   pec15_index;
+    uint16_t pec15 = 16U;
+
+    for (size_t i = 0U; i < size; i++)
+    {
+        pec15_index = ((pec15 >> 7) ^ data[i]) & 0xFF;
+        pec15       = (uint16_t)((pec15 << 8) ^ lookup_tables.crc[pec15_index]);
+    }
+
+    // Set the LSB of the computed PEC15 to 0
+    return (uint16_t)(pec15 << 1);
+}
+
+static ExitCode Io_LTC6813_EnterReadyState(void)
+{
+    uint8_t rx_data;
+    for (size_t i = 0U; i < NUM_OF_LTC6813; i++)
+    {
+        Io_SharedSpi_Receive(ltc_6813.spi, &rx_data, 1U);
+    }
+
+    return EXIT_CODE_OK;
+}
+
+static ExitCode Io_LTC6813_StartADCConversion(void)
+{
+    uint8_t tx_cmd[NUM_OF_CMD_BYTES];
+    tx_cmd[0] = (uint8_t)(ADCV >> 8);
+    tx_cmd[1] = (uint8_t)ADCV;
+
+    uint16_t tx_cmd_pec15 =
+        Io_CalculatePec15(tx_cmd, NUM_OF_PEC15_BYTES_PER_CMD);
+    tx_cmd[2] = (uint8_t)(tx_cmd_pec15 >> 8);
+    tx_cmd[3] = (uint8_t)tx_cmd_pec15;
+
+    return (Io_SharedSpi_Transmit(ltc_6813.spi, tx_cmd, NUM_OF_CMD_BYTES) ==
+            HAL_OK)
+               ? EXIT_CODE_OK
+               : EXIT_CODE_UNIMPLEMENTED;
+}
+
+static ExitCode Io_LTC6813_PollAdcConversion(void)
+{
+    uint8_t tx_cmd[NUM_OF_CMD_BYTES];
+    uint8_t rx_data = 0xFF;
+
+    tx_cmd[0] = (uint8_t)PLADC;
+    tx_cmd[1] = (uint8_t)(PLADC >> 8);
+
+    uint16_t tx_cmd_pec15 =
+        Io_CalculatePec15(tx_cmd, NUM_OF_PEC15_BYTES_PER_CMD);
+    tx_cmd[2] = (uint8_t)(tx_cmd_pec15 >> 8);
+    tx_cmd[3] = (uint8_t)tx_cmd_pec15;
+
+    uint32_t adc_conversion_timeout_counter = 0U;
+
+    while (rx_data == 0xFF)
+    {
+        if (Io_SharedSpi_TransmitAndReceive(
+                ltc_6813.spi, tx_cmd, NUM_OF_CMD_BYTES, &rx_data, 1U) != HAL_OK)
+        {
+            return EXIT_CODE_UNIMPLEMENTED;
+        }
+
+        ++adc_conversion_timeout_counter;
+
+        // Timeout counter threshold of 10 was chosen arbitrarily.
```

Comment:
Tested this function on a daisy chain containing two LTC6813 ICs. Usually takes one cycle in here to finish. The function does not return EXIT_CODE_TIMEOUT unless the power supply has been disconnected from the DC2350A-B board.

Reply:
I'll set an appropriate timeout counter threshold when I test this code with the segment. There are six LTC6813's on the daisy chain so the ADC conversion might take a little longer to complete.